httpserver/handlers.go

```diff
@@ -83,19 +87,96 @@ func (server *Server) verify(ctx context.Context, w http.ResponseWriter, r *http
 				Key:   subject,
 				Value: fromVerifyResult(result),
 			})
+			logrus.Debugf("verification: execution time for image %s: %dms", subject, time.Since(routineStartTime).Milliseconds())
 		}(utils.SanitizeString(subject))
 	}
 	wg.Wait()
+	logrus.Debugf("verification: execution time for request: %dms", time.Since(startTime).Milliseconds())
+	return sendResponse(&results, "", w, http.StatusOK, false)
+}
+
+func (server *Server) mutate(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	startTime := time.Now()
+	sanitizedMethod := utils.SanitizeString(r.Method)
+	sanitizedURL := utils.SanitizeURL(*r.URL)
+	logrus.Infof("start request %s %s", sanitizedMethod, sanitizedURL)
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return fmt.Errorf("unable to read request body: %v", err)
+	}
+
+	// parse request body
+	var providerRequest externaldata.ProviderRequest
+	err = json.Unmarshal(body, &providerRequest)
+	if err != nil {
+		return fmt.Errorf("unable to unmarshal request body: %v", err)
+	}
 
-	return sendResponse(&results, "", w, http.StatusOK)
+	results := make([]externaldata.Item, 0)
+	wg := sync.WaitGroup{}
+	mu := sync.RWMutex{}
+
+	for _, image := range providerRequest.Request.Keys {
+		wg.Add(1)
+		go func(image string) {
+			defer wg.Done()
+			routineStartTime := time.Now()
+			logrus.Infof("image for request %v %v is %v", sanitizedMethod, sanitizedURL, image)
+			returnItem := externaldata.Item{
+				Key:   image,
+				Value: image,
+			}
+			parsedReference, err := pkgUtils.ParseSubjectReference(image)
+			if err != nil {
+				errMessage := fmt.Sprintf("failed to mutate image reference %s: %v", image, err)
+				logrus.Error(errMessage)
+				returnItem.Error = errMessage
+			} else if parsedReference.Digest == "" {
+				var selectedStore referrerstore.ReferrerStore
+				for _, store := range server.GetExecutor().ReferrerStores {
+					if store.Name() == server.MutationStoreName {
+						selectedStore = store
+						break
+					}
+				}
+				if selectedStore == nil {
+					errMessage := fmt.Sprintf("failed to mutate image reference %s: could not find matching store: %v", image, err)
+					logrus.Error(errMessage)
+					returnItem.Error = errMessage
+				} else {
+					descriptor, err := selectedStore.GetSubjectDescriptor(ctx, parsedReference)
+					if err != nil {
+						errMessage := fmt.Sprintf("failed to mutate image reference %s: %v", image, err)
+						logrus.Error(errMessage)
+						returnItem.Error = errMessage
+					} else {
+						returnItem.Value = fmt.Sprintf("%s@%s", parsedReference.Path, descriptor.Digest.String())
+					}
+				}
+			}
```

Comment:
There is multiple levels of nesting here which can make it hard to read. What is the downside of acquiring the lock at start and having the append as part of a defer func? That way you can return on every if condition to avoid the nesting.

Reply:
thanks for the suggestion. I moved the locking and appending functionality to a defer function. I think it makes more sense to only lock once ready to append so that other routines are waiting for the network operation to finish before being able to continue